File: practice/solution/binary-search/tebak_angka.cpp
#include <bits/stdc++.h>
using namespace std;

string guess(int x) {
  printf("%d\n", x);
  fflush(stdout);
  char response[100];
  fgets(response, 100, stdin);
  response[strcspn(response, "\r\n")] = 0; // remove trailing newline
  return response;
}

void solve(int A, int B) {
  while (A <= B) {
    int mid = (A + B) / 2;
    string response = guess(mid);
    if (response == "selamat") {
      return;
    } else if (response == "terlalu kecil") {
      A = mid+1;
    } else if (response == "terlalu besar") {
      B = mid-1;
    } else assert(false);
  }
}

int main() {
  char input[100];
  fgets(input, 100, stdin);
  int A, B;
  sscanf(input, "%d%d", &A, &B);
  solve(A, B);

  return 0;
}

File: practice/solution/binary-search/tebak_angka.go
package main

import (
	"bufio"
	"fmt"
	"io"
	"os"
	"strconv"
	"strings"
)

var out = bufio.NewWriter(os.Stdout)
var in = bufio.NewReader(os.Stdin)

// Guess makes a guess and return "selamat", "terlalu kecil", or "terlalu besar"
func Guess(x int) string {
	if _, err := out.WriteString(fmt.Sprintf("%d\n", x)); err != nil {
		panic(err)
	}
	if err := out.Flush(); err != nil {
		panic(err)
	}

	return getline()
}

func Solve(A, B int) {
	for A <= B {
		m := (A + B) / 2
		res := Guess(m)
		if res == "selamat" {
			return
		} else if res == "terlalu kecil" {
			A = m
		} else if res == "terlalu besar" {
			B = m
		} else {
			panic(fmt.Errorf("unknown response: %s", res))
		}
	}
}

func main() {
	tokens := getints()
	Solve(tokens[0], tokens[1])
}

func getline() string {
	line, _, err := in.ReadLine()
	if err == io.EOF {
		return ""
	}
	if err != nil {
		panic(err)
	}
	return strings.TrimRight(string(line), "\r\n")
}

func getints() []int {
	tokens := strings.Split(getline(), " ")
	ints := make([]int, len(tokens))
	for i, token := range tokens {
		var err error
		ints[i], err = strconv.Atoi(token)
		if err != nil {
			panic(err)
		}
	}
	return ints
}
